README.md
# hARPy

[![](https://img.shields.io/pypi/v/harpy-prjct?label=PyPI)][6]
[![](https://img.shields.io/pypi/pyversions/harpy-prjct?label=Python)][6]
[![](https://img.shields.io/pypi/l/harpy-prjct?label=License)][4]

**hARPy is an active/passive ARP discovery tool. It supports [Python 2][1] (2.7) and [Python 3][1] (3.4 to 3.9) and runs only on GNU/Linux.**

## How It Works

Sends [ARP (Address Resolution Protocol)][2] requests (active mode only) for discovering the link layer addresses and sniffs for ARP replies.

## Features

- Ability to...
    - ...detect suspicious packets during scanning,
    - ...scan active (normal or fast) or passive,
    - ...scan more than one range at the same time,
    - ...filter the results using the given scanning range,
    - ...send packets from a fake IP address,
    - ...show number of hosts and ARP reply/request counts.
- Option to determine...
    - ...the amount of ARP requests to be sent,
    - ...the sleep time between each ARP request.

## Tested OSs

- Kali Linux 2020.4
- Kali Linux 2020.3
- Linux Mint 20 "Ulyana"
- openSUSE Leap 15.2
- Pardus 19.4
- Ubuntu 20.04.1 LTS

## Preparation and Installation

### PyPI

```shell
python3 -m pip install -U pip
python3 -m pip install -U setuptools
python3 -m pip install -U harpy-prjct
```

Note: For Python version 2, change the command "python3" to "python2". But first get the PIP script by doing the following:

```shell
curl https://bootstrap.pypa.io/pip/2.7/get-pip.py -o get-pip.py
python2 get-pip.py
```

### GitHub

```shell
git clone https://github.com/serhatcelik/harpy.git
cd harpy
python3 -m harpy --help
```

## Usage

Use the following command to get [usage][3] help:

```shell
harpy --help
```

## Examples

```
# Active scanning for common IP addresses in fast mode on wlan0
harpy -i wlan0 -f

# Passive scanning
harpy -i wlan0 -p

# Scan a fixed range with a count value of 2
harpy -i wlan0 -r 192.168.0.1/24 -c 2

# Scan some fixed ranges with filtering
harpy -i wlan0 -r 172.16.0.1/16 10.0.0.1/8 -F
```

## Feedback

If you have found a bug or have a suggestion, please consider [creating an issue][5].


[1]: https://www.python.org
[2]: https://en.wikipedia.org/wiki/Address_Resolution_Protocol
[3]: https://github.com/serhatcelik/harpy/wiki#usage
[4]: https://choosealicense.com/licenses/mit/
[5]: https://github.com/serhatcelik/harpy/issues
[6]: https://pypi.org/project/harpy-prjct/


CHANGELOG
########################
# History of the hARPy #
########################

hARPy v2021.9.8
---------------
- It is now checked whether the current operating system is supported or not.
- Code improvements.
- The UI has been improved.
- Documentation has been improved.


hARPy v2021.8.1.post1
---------------------
- Documentation has been improved.


hARPy v2021.8.1
---------------
- Fixed the error "TypeError: 'encoding' is an invalid keyword argument for this function" when trying to install hARPy with Python version 2.7.
- Code improvements.
- Exit statuses have been changed; 0: All is well, 2: Error, 34: Fatal.
- The UI has been improved.
- Documentation has been improved.


hARPy v2021.4.17
----------------
- The UI has been improved.


hARPy v2021.3.10
----------------
- Code improvements.
- Performance improvements.
- The UI has been improved.


hARPy v2021.1.13
----------------
- The UI has been improved.


hARPy v2021.1.12
----------------
- Fixed an error where unknown MAC addresses could not be determined as "unknown".


hARPy v2021.1.9
---------------
- Fixed a memory error that occurred when the sniff results container was full.
- Exit codes have been made more meaningful.


hARPy v2021.1.2
---------------
- Fixed an error where an unavailable interface could be selected.
- Fixed an error where the socket could not be closed after a critical error.
- Fixed an error that caused the program to crash when an invalid packet was caught.
- Fixed an error that caused the vendor text length to be measured incorrectly.
- Fixed an error that occurred when trying to print a message to an inactive terminal.
- Performance improvements.
- Added Python version 2.7 support.
- Repeat mode function has been added.
- OUI database has been updated.
- Different vendor information about suspicious packets is now displayed.
- Superuser control is now done with geteuid instead of getuid.
- Repetitive scanning ranges are now counted as one.
- Non-ASCII characters are no longer printed.
- Exit codes have been made more meaningful.
- The UI has been improved.
- Type less, do more.


hARPy v2020.11.21.post1
-----------------------
- Minor bug fixes.


hARPy v2020.11.21
-----------------
- Fixed the "Set changed size during iteration" error.
- Fixed an error when trying to get the length of an inactive terminal.
- Fixed an error that occurred when trying to close a socket that does not exist.
- Performance improvements.
- Added Python 3.4 and Python 3.5 support.
- Fast mode function has been added.
- Scanning range can now take multiple values.
- Common IP addresses are now automatically scanned when no scanning range is specified.
- The UI has been improved.
- Type less, do more.


hARPy v2020.11.3.post1
----------------------
- Minor bug fixes.


hARPy v2020.11.3
----------------
- Fixed an error where the filter is invalid due to incorrectly entered scanning range.
- Performance improvements.
- The UI has been improved.


hARPy v2020.11.1
----------------
- Fixed an error that occurred when trying to run the program with no internet connection.
- A command-line option that filters sniff results has been added.
- MAC addresses of suspicious packets are now only shown in one column.
- The UI has been improved.
- Type less, do more.


hARPy v2020.10.30.post1
-----------------------
- Performance improvements.


hARPy v2020.10.30
-----------------
- Fixed an error that occurred when running the program in the background.
- Fixed an error that occurred when running the program outside of the terminal.
- The UI has been improved.


hARPy v2020.10.27.post1
-----------------------
- Minor bug fixes.


hARPy v2020.10.27
-----------------
- Fixed an error that caused the processes to continue after death of the terminal.


hARPy v2020.10.26
-----------------
- Fixed an error that caused the program to wait too long before exit.
- Timeout function has been added.
- The UI has been improved.


hARPy v2020.10.23.post1
-----------------------
- Minor bug fixes.


hARPy v2020.10.23
-----------------
- Fixed an error that caused the program to crash on exit.
- Fixed an error that caused scan results to be lost on exit.
- Fixed an error that occurred when encountering an unexpected socket closure.
- A command-line option that shows error logs has been added.
- A command-line option that shows program author information has been added.
- The UI has been improved.


hARPy v2020.9.25
----------------
- The error that occurred when encountering signal 20 has been fixed.


hARPy v2020.9.17
----------------
- First version.



2020.7.19
---------
I have an idea!


harpy/handlers.py
# coding=utf-8

# This file is part of hARPy
# Released under the MIT license
# Copyright (C) Serhat Çelik

r"""
EXAMPLES
  1) Active scanning for common IP addresses in fast mode on wlan0
  harpy -i wlan0 -f

  2) Passive scanning
  harpy -i wlan0 -p

  3) Scan a fixed range with a count value of 2
  harpy -i wlan0 -r 192.168.0.1/24 -c 2

  4) Scan some fixed ranges with filtering
  harpy -i wlan0 -r 172.16.0.1/16 10.0.0.1/8 -F

ABOUT
  It is recommended to enable passive mode on networks with heavy packet flow.
  See https://github.com/serhatcelik/harpy for more information.
"""

from __future__ import print_function

import argparse
import binascii
import io
import json
import os
import re
import signal
import socket
import struct
import subprocess
import sys
import termios
import threading

from harpy import notice, data
from harpy.data import get_logo, get_banner, add_colons, add_dots, run_main


class ExceptionHandler(object):
    def __init__(self, who=None):
        self.who = who  # Responsible for the error

    def __call__(self, func):
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except (IOError, OSError, socket.error, termios.error) as err:
                # 5: Input/output error
                if err.args[0] == 5:
                    # Mostly for "print" errors
                    pass
                elif err.args[0] in (6, 9, 19, 100):
                    # 6: No such device or address
                    # 9: Bad file descriptor
                    # 19: No such device
                    # 100: Network is down
                    self.add_exception(err.args[0], err.args[1])
                else:
                    raise

            return run_main(False)

        return wrapper

    def add_exception(self, erno, error):
        """
        Update the exit messages container with a new error message.

        :param erno: Error number.
        :param error: Error content.
        """

        data.EXIT_MSGS.add("[!] %s -> [Errno %d] %s" % (self.who, erno, error))


class ArgumentHandler(object):
    def __init__(self):
        pass

    @staticmethod
    def handle_count(count):
        if count < data.MIN_CNT:
            data.CNT = data.MIN_CNT

    @staticmethod
    @ExceptionHandler()
    def handle_interface(interface):
        if interface is False:
            print("[!] No available interface in %s" % data.SYS_PATH)
            sys.stdout.flush()
            return False
        if interface == "lo":
            print("[!] '%s': This is not an Ethernet interface" % interface)
            sys.stdout.flush()
            return False
        if interface not in InterfaceHandler().members:
            print("[!] '%s': No such interface" % interface)
            sys.stdout.flush()
            return False
        if InterfaceHandler().members[interface] != "up":
            print("[!] '%s': Interface not available (dormant?)" % interface)
            sys.stdout.flush()
            return False
        return True

    @staticmethod
    def handle_log():
        if os.path.isfile(data.LOG_FILE):
            with io.open(data.LOG_FILE, "r", encoding="utf-8") as log:
                return log.read()
        return "No log"

    @staticmethod
    def handle_node(node):
        if not data.MIN_NOD <= node <= data.MAX_NOD:
            data.NOD = data.DEF_NOD

    @staticmethod
    def handle_passive(passive):
        if passive:
            # Fast mode only makes sense in active mode, so...
            data.FST = None

    @staticmethod
    @ExceptionHandler()
    def handle_range(range_):
        if range_ is None:
            # Filtering is only allowed if a scanning range is specified, so...
            data.FLT = None
            # Repeat is only allowed if a scanning range is specified, so...
            data.REP = None
            range_ = data.RNG = data.DEF_RNG

        # Do this only if the scanning range is different from the default
        if range_ != data.DEF_RNG:
            octet = "([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])"
            pattern = r"^({0}\.{0}\.{0}\.{0}/(8|16|24))$".format(octet)

            set_range_ = []
            # Remove repetitive scanning ranges
            for _ in range_:
                if _ not in set_range_:
                    set_range_.append(_)
            range_ = set_range_

            problem = [_ for _ in range_ if not bool(re.match(pattern, _))]
            if problem:
                print("Problem with scanning range(s):", end=" ")
                for i, _ in enumerate(problem):
                    # Last error?
                    if i == len(problem) - 1:
                        print(_)
                    else:
                        print(_, end=data.SEPARATOR)
                sys.stdout.flush()
                return False

        data.RNG = [[
            _.split(".")[0],  # pylint: disable=C0207
            _.split(".")[1],
            _.split(".")[2],
            _.split(".")[3].split("/")[0],
            _.split(".")[3].split("/")[-1],
        ] for _ in range_]  # Convert the scanning range to list format

        return True

    @staticmethod
    def handle_sleep(sleep):
        if sleep < data.MIN_SLP:
            data.SLP = data.MIN_SLP
        elif sleep > data.MAX_SLP:
            data.SLP = data.MAX_SLP

    @staticmethod
    def handle_timeout(timeout):
        if timeout < data.MIN_TIM:
            data.TIM = data.MIN_TIM


class EchoHandler(object):
    def __init__(self):
        self.descriptor = sys.stdin.fileno()

    @ExceptionHandler()
    def enable(self):
        """
        Enable terminal echo.
        """

        new = termios.tcgetattr(self.descriptor)
        new[3] |= termios.ECHO
        termios.tcsetattr(self.descriptor, termios.TCSANOW, new)

    @ExceptionHandler()
    def disable(self):
        """
        Disable terminal echo.
        """

        new = termios.tcgetattr(self.descriptor)
        new[3] &= ~termios.ECHO
        termios.tcsetattr(self.descriptor, termios.TCSANOW, new)


class InterfaceHandler(object):
    members = {}  # All interfaces

    def __init__(self):
        if hasattr(socket, "if_nameindex"):
            if_nameindex = socket.if_nameindex()  # pylint: disable=E1101
            self.members = {_[-1]: None for _ in if_nameindex}
        else:
            if os.path.isdir(data.SYS_PATH):
                self.members = {_: None for _ in os.listdir(data.SYS_PATH)}

        for _ in self.members:
            if _ != "lo":
                operstate_file = os.path.join(data.SYS_PATH, _, "operstate")
                if os.path.isfile(operstate_file):
                    with io.open(operstate_file, "r", encoding="utf-8") as opf:
                        try:
                            self.members[_] = opf.read().strip().lower()
                        except (IOError, OSError) as err:
                            # 22: Invalid argument
                            if err.args[0] == 22:
                                pass
                            else:
                                raise

    def __call__(self):
        for _ in self.members:
            if self.members[_] == "up":
                return _
        return False

    @staticmethod
    def get_mac(l2soc):
        """
        Return the MAC address of an interface.

        :param l2soc: Layer 2 RAW socket.
        """

        return binascii.hexlify(l2soc.getsockname()[-1]).decode("utf-8")


class PacketHandler(object):
    def __init__(self):
        pass

    @staticmethod
    def create_eth_frame():
        return struct.pack(
            "!6s6s2s",
            binascii.unhexlify(data.DST_MAC),
            binascii.unhexlify(data.SRC_MAC),
            binascii.unhexlify(data.ETH_TYP),
        )

    @staticmethod
    def create_arp_header():
        return struct.pack(
            "!2s2s1s1s2s6s4s6s4s",
            binascii.unhexlify(data.ARP_HWT),
            binascii.unhexlify(data.ARP_PRT),
            binascii.unhexlify(data.ARP_HWS),
            binascii.unhexlify(data.ARP_PRS),
            binascii.unhexlify(data.ARP_REQ),
            binascii.unhexlify(data.SND_MAC),
            socket.inet_aton(data.SND_IP),
            binascii.unhexlify(data.TGT_MAC),
            socket.inet_aton(data.TGT_IP),
        )


class ParserHandler(object):
    def __init__(self):
        pass

    @staticmethod
    def create_arguments():
        parser = argparse.ArgumentParser(
            prog="harpy", usage="%(prog)s [<extra_option> ...]",
            description="hARPy - Active/passive ARP discovery tool\n"
                        "Written by %s "
                        "(with the help of my family and a friend)"
                        "" % notice.AUTHOR,
            epilog=__doc__, formatter_class=argparse.RawTextHelpFormatter,
        )

        ######################
        # Optional Arguments #
        ######################
        parser.add_argument(
            "-c", metavar="<count>", dest="c", action="store",
            default=data.DEF_CNT, type=int,
            help="send each request %%(metavar)s times, see example 3 "
                 "[DEFAULT:%%(default)s|MINIMUM:%d]" % data.MIN_CNT,
        )
        parser.add_argument(
            "-f", "--fast", dest="f", action="store_true", default=False,
            help="enable fast mode, only scan specific hosts, see example 1",
        )
        parser.add_argument(
            "-F", "--filter", dest="F", action="store_true", default=False,
            help="filter results using given scanning range, see example 4",
        )
        parser.add_argument(
            "-i", metavar="<interface>", dest="i", action="store",
            default=InterfaceHandler()(),
            help="use %(metavar)s as net device to send/sniff, see example 1",
        )
        parser.add_argument(
            "-L", "--license", dest="license", action="version", default=None,
            version=notice.__doc__, help="show license and exit",
        )
        parser.add_argument(
            "-l", "--log", dest="log", action="version", default=None,
            version=ArgumentHandler.handle_log(), help="show log and exit",
        )
        parser.add_argument(
            "-n", metavar="<node>", dest="n", action="store",
            default=data.DEF_NOD, type=int,
            help="use %%(metavar)s as last ip octet to send packets "
                 "[DEFAULT:%%(default)s|MIN:%d|MAX:%d]" % (data.MIN_NOD,
                                                           data.MAX_NOD),
        )
        parser.add_argument(
            "-p", "--passive", dest="p", action="store_true", default=False,
            help="enable passive mode, do not send any packets, see example 2",
        )
        parser.add_argument(
            "-R", "--repeat", dest="R", action="store_true", default=False,
            help="enable repeat mode, never stop sending packets"
        )
        parser.add_argument(
            "-r", metavar="<rng>", dest="r", action="store", nargs="+",
            default=None, help="use %(metavar)s as scanning range",
        )
        parser.add_argument(
            "-s", metavar="<time>", dest="s", action="store",
            default=data.DEF_SLP, type=int,
            help="sleep %%(metavar)s milliseconds between each request "
                 "[DEFAULT:%%(default)s|MIN:%d|MAX:%d]" % (data.MIN_SLP,
                                                           data.MAX_SLP),
        )
        parser.add_argument(
            "-t", metavar="<timeout>", dest="t", action="store",
            default=data.DEF_TIM, type=int,
            help="stop scanning after %%(metavar)s seconds "
                 "[DEFAULT:%%(default)s|MIN:%d]" % data.MIN_TIM,
        )
        parser.add_argument(
            "-v", "--version", dest="version", action="version", default=None,
            version="v" + notice.VERSION, help="show program version and exit",
        )

        return parser.parse_args()

    @staticmethod
    def create_links(commands):
        """
        Create shortcuts to the commands.

        :param commands: Parsed command-line arguments.
        """

        data.CNT = commands.c
        data.FST = commands.f
        data.FLT = commands.F
        data.INT = commands.i
        data.NOD = commands.n
        data.PAS = commands.p
        data.RNG = commands.r
        data.REP = commands.R
        data.SLP = commands.s
        data.TIM = commands.t

    @staticmethod
    def check_arguments():
        return False not in [
            ArgumentHandler.handle_count(data.CNT),
            ArgumentHandler.handle_interface(data.INT),
            ArgumentHandler.handle_node(data.NOD),
            ArgumentHandler.handle_passive(data.PAS),
            ArgumentHandler.handle_range(data.RNG),
            ArgumentHandler.handle_sleep(data.SLP),
            ArgumentHandler.handle_timeout(data.TIM),
        ]


class ResultHandler(object):
    snd_ip = None
    src_mac = None
    snd_mac = None
    arp_opc = None

    def __init__(self):
        self.ouis = self.open_ouis()  # Get OUIs database

    def __call__(self, results):
        for _ in range(0, len(results), data.CONT_STP_SIZ):
            if (self.snd_ip in results) and (self.snd_ip == results[_]):
                if self.src_mac == results[_ + 1]:
                    if self.snd_mac == results[_ + 2]:
                        if self.arp_opc != data.ARP_REQ:
                            results[_ + 3] += 1
                        else:
                            results[_ + 4] += 1
                        return results

        results.append(self.snd_ip)  # Sender IP address
        results.append(self.src_mac)  # Source MAC address
        results.append(self.snd_mac)  # Sender MAC address
        results.append(1 if self.arp_opc != data.ARP_REQ else 0)  # Reply
        results.append(1 if self.arp_opc == data.ARP_REQ else 0)  # Request
        results.append(self.get_vendor(self.src_mac))  # Ethernet vendor
        results.append(self.get_vendor(self.snd_mac))  # ARP vendor

        return results

    @staticmethod
    def open_ouis():
        """
        Obtain the contents of the file that contains OUIs.
        """

        ouis_file = os.path.join(os.path.dirname(__file__), "ouis.json")
        if os.path.isfile(ouis_file):
            with io.open(ouis_file, "r", encoding="utf-8") as ouis:
                try:
                    return json.load(ouis)
                except ValueError:
                    pass
        return None

    def get_vendor(self, mac):
        """
        Find a vendor using the given MAC address.

        :param mac: MAC address to find the vendor.
        """

        if self.ouis is not None:
            if mac[:6] in self.ouis:
                # Prevent the "ordinal not in range(128)" error in 2.7
                # Also prevent for incorrect measurement of text length
                return self.ouis[mac[:6]].encode("ascii", "ignore").decode()
            return "unknown"
        return ""


class SignalHandler(object):
    def __init__(self):
        self.main_thread = vars(threading)["_MainThread"]

    def __call__(self, _signum, _frame):
        data.EXIT_MSGS.add("[x] Exiting, received signal %d" % _signum)
        run_main(False)

    def catch(self, *signals):
        if isinstance(threading.current_thread(), self.main_thread):
            for _ in signals:
                signal.signal(_, self.__call__)

    def ignore(self, *signals):
        # Be sure to ignore the signals
        while isinstance(threading.current_thread(), self.main_thread):
            try:
                for _ in signals:
                    signal.signal(_, signal.SIG_IGN)
                return
            except TypeError:
                # Workaround for _thread.interrupt_main bug from:
                #   https://bugs.python.org/issue23395
                pass


class SocketHandler(object):
    def __init__(self, protocol):
        self.l2soc = socket.socket(
            socket.PF_PACKET, socket.SOCK_RAW, socket.htons(protocol)
        )  # Open a socket

    def set_options(self):
        self.l2soc.setblocking(False)  # Non-blocking mode
        self.l2soc.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.l2soc.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)

    @ExceptionHandler(data.SOCKET)
    def bind(self, interface, port):
        """
        Bind the socket to an interface.

        :param interface: Network device to send/sniff packets.
        :param port: Port to bind to an interface.
        """

        self.l2soc.bind((interface, port))

    def close(self):
        self.l2soc.close()


class WindowHandler(object):
    logo = get_logo()

    def __init__(self, results):
        self.results = results

        self.col_length = self.get_column_length()

        self.banner = get_banner()
        self.banner_results = [
            len(list(_ for _ in range(0, len(results), data.CONT_STP_SIZ))),
            sum(results[_] for _ in range(3, len(results), data.CONT_STP_SIZ)),
            sum(results[_] for _ in range(4, len(results), data.CONT_STP_SIZ)),
        ]

        for i, _ in enumerate(self.banner_results):
            self.banner[i] += str(_)

    @ExceptionHandler()
    def __call__(self):
        for _ in range(0, len(self.results), data.CONT_STP_SIZ):
            ip_address = self.results[_]
            eth_mac_address = add_colons(self.results[_ + 1])
            arp_mac_address = add_colons(self.results[_ + 2])
            arp_rep = str(self.results[_ + 3])
            rep_space_len = data.MAX_REP_LEN - len(arp_rep)
            # Prevent column distortion
            arp_rep = str(float("inf")) if rep_space_len < 0 else arp_rep
            arp_req = str(self.results[_ + 4])
            req_space_len = data.MAX_REQ_LEN - len(arp_req)
            arp_req = str(float("inf")) if req_space_len < 0 else arp_req
            eth_vendor = self.results[_ + 5]
            arp_vendor = self.results[_ + 6]

            # Suspicious packet?!
            if eth_mac_address != arp_mac_address:
                if data.ETHER_TO_ARP:
                    mac_address = arp_mac_address + "!"
                    vendor = arp_vendor
                else:
                    mac_address = eth_mac_address + "!"
                    vendor = eth_vendor
            else:
                mac_address = eth_mac_address
                vendor = eth_vendor

            print(ip_address.ljust(data.MAX_IP_LEN), end=data.SEPARATOR)
            print(mac_address.ljust(data.MAX_MAC_LEN), end=data.SEPARATOR)
            print(arp_rep.ljust(data.MAX_REP_LEN), end=data.SEPARATOR)
            print(arp_req.ljust(data.MAX_REQ_LEN), end=data.SEPARATOR)
            vendor = add_dots(vendor, self.col_length, data.MAX_ALL_LEN)
            print(vendor)
            sys.stdout.flush()

    @staticmethod
    @ExceptionHandler()
    def get_column_length():
        if hasattr(os, "get_terminal_size"):
            return os.get_terminal_size().columns  # pylint: disable=E1101
        return int(subprocess.check_output(["tput", "cols"]).strip())

    @ExceptionHandler()
    def draw_a_line(self):
        print("-" * self.col_length)
        sys.stdout.flush()

    @staticmethod
    @ExceptionHandler()
    def draw_a_row(*args):
        """
        Draw a row for the result window.

        :param args: Container that stores column texts.
        """

        for i, _ in enumerate(args):
            # Last column?
            if i == len(args) - 1:
                print(_)
                sys.stdout.flush()
            else:
                print(_, end=data.SEPARATOR)

    @ExceptionHandler()
    def draw_skeleton(self):
        #################
        # Logo & Banner #
        #################
        for i, j in zip(self.logo, self.banner):
            print(i + (" " * (data.MAX_IP_LEN - len(i))), end=data.SEPARATOR)
            print(j)
        sys.stdout.flush()

        ######################
        # MAC Address Column #
        ######################
        data.ETHER_TO_ARP = not data.ETHER_TO_ARP

        ########
        # Rows #
        ########
        if data.TGT_IP is None:
            info_col = "Sending disabled"
        elif data.TGT_IP is False:
            info_col = "Sending finished"
        else:
            info_col = "Sending .%d -> %s" % (data.NOD, data.TGT_IP)
            if data.REP:
                info_col = "R/" + info_col

        self.draw_a_line()
        self.draw_a_row("Ctrl+C to exit".ljust(data.MAX_IP_LEN), info_col)
        self.draw_a_line()
        self.draw_a_row(
            "IP Address".ljust(data.MAX_IP_LEN),
            "MAC Address".ljust(data.MAX_MAC_LEN),
            "Reply".ljust(data.MAX_REP_LEN),
            "Request".ljust(data.MAX_REQ_LEN),
            "Vendor",
        )
        self.draw_a_line()
